File: backend/app/api/routes/pasivos.py
from __future__ import annotations

from decimal import Decimal
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.db.models import Moneda, PasivoEstado
from app.db.session import get_db
from app.schemas.pasivos import (
    CancelarAcreedorChequeRequest,
    PagoAcreedorRequest,
    PagoAcreedorResponse,
    PasivoCancelarConChequeRequest,
    PasivoCreate,
    PasivoImputadoRead,
    PasivoPagoRequest,
    PasivoRead,
    PasivoUpdate,
)
from app.services import cheques as svc_cheques
from app.services import pasivos as service

router = APIRouter(prefix="/pasivos", tags=["pasivos"])

DbSession = Annotated[Session, Depends(get_db)]


@router.post("", response_model=PasivoRead, status_code=201)
def create_pasivo(payload: PasivoCreate, db: DbSession) -> PasivoRead:
    return service.create_pasivo(db, payload)


@router.get("", response_model=list[PasivoRead])
def list_pasivos(
    db: DbSession,
    estado: PasivoEstado | None = None,
) -> list[PasivoRead]:
    return service.list_pasivos(db, estado=estado)


def _respuesta_acreedor(
    resultado: service.PagoAcreedorResult, moneda_deuda: Moneda
) -> PagoAcreedorResponse:
    """Arma la respuesta HTTP a partir del resultado del servicio.

    El servicio devuelve los modelos vivos (los usa el bot para redactar su
    mensaje); acá se recorta a lo que el panel necesita mostrar."""
    return PagoAcreedorResponse(
        acreedor=resultado.acreedor,
        moneda_deuda=moneda_deuda,
        imputaciones=[
            PasivoImputadoRead(
                id=imp.pasivo.id,
                concepto=imp.pasivo.concepto,
                imputado=imp.imputado,
                saldo_restante=imp.pasivo.saldo_pendiente,
                cancelo=imp.cancelo,
            )
            for imp in resultado.imputaciones
        ],
        imputado=sum((imp.imputado for imp in resultado.imputaciones), Decimal("0.00")),
        saldo_restante=resultado.saldo_restante,
        cancelados=resultado.cancelados,
    )


@router.post("/acreedores/pagar", response_model=PagoAcreedorResponse)
def pagar_a_acreedor(
    payload: PagoAcreedorRequest, db: DbSession
) -> PagoAcreedorResponse:
    """Paga todas las deudas con un acreedor, de la más vieja a la más nueva.

    Es la misma función que usa el bot: el reparto no depende de por dónde entró
    la orden."""
    resultado = service.pagar_a_acreedor(
        db,
        acreedor=payload.acreedor,
        moneda_deuda=payload.moneda_deuda,
        monto_pagado=payload.monto_pagado,
        moneda_pago=payload.moneda_pago,
        medio_pago=payload.medio_pago,
        cotizacion=payload.cotizacion,
        fecha=payload.fecha,
    )
    return _respuesta_acreedor(resultado, payload.moneda_deuda)


@router.post("/acreedores/cancelar-con-cheque", response_model=PagoAcreedorResponse)
def cancelar_a_acreedor_con_cheque(
    payload: CancelarAcreedorChequeRequest, db: DbSession
) -> PagoAcreedorResponse:
    """Entrega un cheque de cartera contra todas las deudas en pesos del acreedor.

    Un cheque es un instrumento en pesos: no hay con qué imputarlo a una deuda en
    dólares sin una cotización que nadie dictó."""
    cheque = svc_cheques.get_cheque(db, payload.cheque_id)
    resultado = service.cancelar_a_acreedor_con_cheque(
        db,
        acreedor=payload.acreedor,
        cheque=cheque,
        porcentaje_venta=payload.porcentaje_venta,
        operador_id=payload.operador_id,
        motivo=payload.motivo,
        fecha=payload.fecha,
        vuelto_modo=payload.vuelto_modo,
    )
    return _respuesta_acreedor(resultado, Moneda.ARS)


@router.get("/{pasivo_id}", response_model=PasivoRead)
def get_pasivo(pasivo_id: UUID, db: DbSession) -> PasivoRead:
    return service.get_pasivo(db, pasivo_id)


@router.patch("/{pasivo_id}", response_model=PasivoRead)
def editar_pasivo(pasivo_id: UUID, payload: PasivoUpdate, db: DbSession) -> PasivoRead:
    return service.editar_pasivo(db, pasivo_id, payload)


@router.post("/{pasivo_id}/pagar", response_model=PasivoRead)
def pagar_pasivo(
    pasivo_id: UUID,
    payload: PasivoPagoRequest,
    db: DbSession,
) -> PasivoRead:
    return service.pagar_pasivo(db, pasivo_id, payload)


@router.post("/{pasivo_id}/cancelar-con-cheque", response_model=PasivoRead)
def cancelar_pasivo_con_cheque(
    pasivo_id: UUID,
    payload: PasivoCancelarConChequeRequest,
    db: DbSession,
) -> PasivoRead:
    return service.cancelar_con_cheque(db, pasivo_id, payload)

File: frontend/src/api/pasivos.ts
import { apiFetch } from './client'
import type { MedioPago, Moneda, Pasivo, PasivoEstado } from '../types'

export interface PasivoCreatePayload {
  acreedor: string
  concepto: string
  monto: number
  moneda: Moneda
  fecha_vencimiento: string | null
  observaciones: string | null
  // Solo cuando le prestaron plata al negocio: además de anotar la deuda, asienta
  // el ingreso del efectivo que entró al cajón.
  ingreso_caja?: boolean
  fecha_ingreso?: string | null
  // Obligatoria si le prestaron DÓLARES: costo ($/USD) del lote de stock que se
  // crea, sin el cual esos dólares no se pueden vender.
  cotizacion_ingreso_usd?: number | null
}

export const getPasivos = (estado?: PasivoEstado): Promise<Pasivo[]> => {
  const qs = estado ? `?estado=${estado}` : ''
  return apiFetch<Pasivo[]>(`/pasivos${qs}`)
}

export const createPasivo = (payload: PasivoCreatePayload): Promise<Pasivo> =>
  apiFetch<Pasivo>('/pasivos', { method: 'POST', body: JSON.stringify(payload) })

// Corrección de la carga de una deuda. `monto`/`moneda` solo si está PENDIENTE y
// sin pagos parciales (lo valida el backend).
export interface PasivoUpdatePayload {
  acreedor?: string
  concepto?: string
  monto?: number
  moneda?: Moneda
  fecha_vencimiento?: string | null
  observaciones?: string | null
  ingreso_caja?: boolean
  fecha_ingreso?: string | null
  cotizacion_ingreso_usd?: number | null
}

export const editarPasivo = (id: string, payload: PasivoUpdatePayload): Promise<Pasivo> =>
  apiFetch<Pasivo>(`/pasivos/${id}`, { method: 'PATCH', body: JSON.stringify(payload) })

export interface PagarPasivoPayload {
  monto_pagado: number
  moneda_pago: Moneda
  medio_pago: MedioPago
  // Requerida solo si moneda_pago difiere de la moneda de la deuda ($/USD).
  cotizacion?: number | null
  fecha_cancelacion?: string | null
}

export const pagarPasivo = (id: string, payload: PagarPasivoPayload): Promise<Pasivo> =>
  apiFetch<Pasivo>(`/pasivos/${id}/pagar`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

export type VueltoModo = 'SALDAR_EFECTIVO' | 'QUEDA_DEBIENDO'

export interface CancelarConChequePayload {
  cheque_id: string
  porcentaje_venta: number
  operador_id: string
  motivo: string
  fecha_cancelacion?: string | null
  // Requerido solo si el cheque cubre de más (diferencia > 0).
  vuelto_modo?: VueltoModo | null
}

export const cancelarPasivoConCheque = (
  id: string,
  payload: CancelarConChequePayload,
): Promise<Pasivo> =>
  apiFetch<Pasivo>(`/pasivos/${id}/cancelar-con-cheque`, {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Pago consolidado a un acreedor — el espejo del cobro por cliente (§2.c).
 *
 * El negocio le puede deber varias veces a la misma persona: le compró un lote
 * de dólares sin pagarlo, después un cheque, y encima le pidió plata prestada.
 * Cuando le paga **no está pagando una de esas deudas: está pagando lo que le
 * debe**. El importe se imputa de la deuda más vieja a la más nueva y todo entra
 * en una sola transacción, con su propia línea de caja por deuda.
 *
 * El acreedor es **texto libre** (no un cliente con id) y el backend lo matchea
 * exacto, sin distinguir mayúsculas. Agrupá con el mismo criterio o la pantalla
 * va a mostrar un total y el botón va a pagar otro.
 */

/** Una deuda alcanzada por el pago, y cuánto le tocó. */
export interface PasivoImputado {
  id: string
  concepto: string
  imputado: string
  saldo_restante: string
  cancelo: boolean
}

export interface PagoAcreedorResult {
  acreedor: string
  moneda_deuda: Moneda
  /** Las deudas alcanzadas, de la más vieja a la más nueva. */
  imputaciones: PasivoImputado[]
  /** Cuánto bajó la deuda en total, en `moneda_deuda`. */
  imputado: string
  saldo_restante: string
  cancelados: number
}

export interface PagarAcreedorPayload {
  acreedor: string
  /** Contra cuál de las dos colas se imputa: ARS y USD no se suman. */
  moneda_deuda: Moneda
  monto_pagado: number
  moneda_pago: Moneda
  medio_pago: MedioPago
  // Requerida solo si moneda_pago difiere de moneda_deuda ($/USD).
  cotizacion?: number | null
  fecha?: string | null
}

export const pagarAcreedor = (payload: PagarAcreedorPayload): Promise<PagoAcreedorResult> =>
  apiFetch<PagoAcreedorResult>('/pasivos/acreedores/pagar', {
    method: 'POST',
    body: JSON.stringify(payload),
  })

/**
 * Entrega un cheque de cartera contra todas las deudas **en pesos** del acreedor.
 *
 * El cheque vale su neto (nominal menos el descuento pactado) y con eso se van
 * llenando las deudas, de la más vieja a la más nueva. No mueve efectivo: el
 * desembolso ocurrió al comprar el cheque, acá solo cambia de manos el papel.
 * Un cheque es un instrumento en pesos, así que las deudas en dólares no entran.
 */
export interface CancelarAcreedorChequePayload {
  acreedor: string
  cheque_id: string
  porcentaje_venta: number
  operador_id: string
  motivo: string
  fecha?: string | null
  // Requerido solo si el neto del cheque cubre de más.
  vuelto_modo?: VueltoModo | null
}

export const cancelarAcreedorConCheque = (
  payload: CancelarAcreedorChequePayload,
): Promise<PagoAcreedorResult> =>
  apiFetch<PagoAcreedorResult>('/pasivos/acreedores/cancelar-con-cheque', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
